mockwebserver/src/main/java/okhttp3/mockwebserver/RecordedRequest.kt

```diff
@@ -14,154 +14,102 @@
+   * was empty or unchunked.
+   */
+  val chunkSizes: List<Int>?,
+  /** The total size of the body of this POST request (before truncation).*/
+  val bodySize: Long,
+  /** The body of this POST request. This may be truncated. */
+  val body: Buffer?,
+  /**
+   * The index of this request on its HTTP connection. Since a single HTTP connection may
+   * serve multiple requests, each request is assigned its own sequence number.
+   */
+  val sequenceNumber: Int,
+  socket: Socket
+) {
+  val method: String?
+  val path: String?
+  /**
+   * The TLS handshake of the connection that carried this request, or null if the request
+   * was received without TLS.
+   */
+  val handshake: Handshake?
+  val requestUrl: HttpUrl?
 
-import java.io.IOException;
-import java.net.Inet6Address;
-import java.net.InetAddress;
-import java.net.Socket;
-import java.util.List;
-import javax.net.ssl.SSLSocket;
-import okhttp3.Handshake;
-import okhttp3.Headers;
-import okhttp3.HttpUrl;
-import okhttp3.TlsVersion;
-import okio.Buffer;
+  val utf8Body: String?
+    @Deprecated("Use {@link #getBody() getBody().readUtf8()}. ")
+    get() = body?.readUtf8()
 
-/** An HTTP request that came into the mock web server. */
-public final class RecordedRequest {
-  private final String requestLine;
-  private final String method;
-  private final String path;
-  private final Headers headers;
-  private final Handshake handshake;
-  private final List<Integer> chunkSizes;
-  private final long bodySize;
-  private final Buffer body;
-  private final int sequenceNumber;
-  private final HttpUrl requestUrl;
+  /** Returns the connection's TLS version or null if the connection doesn't use SSL.  */
+  val tlsVersion: TlsVersion?
+    get() = handshake?.tlsVersion()
 
-  public RecordedRequest(String requestLine, Headers headers, List<Integer> chunkSizes,
-      long bodySize, Buffer body, int sequenceNumber, Socket socket) {
-    this.requestLine = requestLine;
-    this.headers = headers;
-    this.chunkSizes = chunkSizes;
-    this.bodySize = bodySize;
-    this.body = body;
-    this.sequenceNumber = sequenceNumber;
-    if (socket instanceof SSLSocket) {
+  init {
+    if (socket is SSLSocket) {
       try {
-        this.handshake = Handshake.get(((SSLSocket) socket).getSession());
-      } catch (IOException e) {
-        throw new IllegalArgumentException(e);
+        this.handshake = Handshake.get(socket.session)
+      } catch (e: IOException) {
+        throw IllegalArgumentException(e)
       }
     } else {
-      this.handshake = null;
+      this.handshake = null
     }
 
-    if (requestLine != null) {
-      int methodEnd = requestLine.indexOf(' ');
-      int pathEnd = requestLine.indexOf(' ', methodEnd + 1);
-      this.method = requestLine.substring(0, methodEnd);
-      String path = requestLine.substring(methodEnd + 1, pathEnd);
+    if (requestLine != null && requestLine.isNotEmpty()) {
```

Comment:
why add the isNotEmpty() check?

Reply:
If requestLine is empty && requestLine !=null
Refer to:
```
      val methodEnd = requestLine.indexOf(' ')
      val pathEnd = requestLine.indexOf(' ', methodEnd + 1)
      this.method = requestLine.substring(0, methodEnd)
```
`methodEnd` becomes -1  for ' ' is not found in empty string,
`this.method = requestLine.substring(0, methodEnd)`,this  line will cause `StringIndexOutOfBoundsException: String index out of range: -1`

Reply:
That's okay. The correct behavior is a crash.

Reply:
After the `val requestLine` is declared as a `non-nullable`, 
The `RecordedRequest` is initialized like this:
`recordedRequest = RecordedRequest(requestLine:"", Headers.of(), listOf(), 0L, Buffer(), 0, Socket())`
Instead of:
`RecordedRequest(requestLine:null, null, null, 0L, null, 0, Socket())`

So use `if requestLine.isNotEmpty()` insteal of `if requestLine!=null`. 
If this decision is removed, some tests will failed such as the test method `recoverFromConnectionNoNewStreamsOnFollowUp` of the `HttpOverHttp2Test` class. 
```
Exception in thread "OkHttp Http2Connection" java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.substring(String.java:1967)
	at okhttp3.mockwebserver.RecordedRequest.<init>(RecordedRequest.kt:81)
	at okhttp3.mockwebserver.MockWebServer.dispatchBookkeepingRequest(MockWebServer.java:634)
```